triton-constraint-circuit/src/lib.rs

```diff
@@ -1471,4 +1705,316 @@ mod tests {
+                    main_input.view(),
+                    aux_input.view(),
+                    &challenges,
+                )
+            })
+            .collect_vec();
+
+        // apply one step of degree-lowering
+        let num_nodes = multicircuit_monad[0].builder.all_nodes.borrow().len();
+        let (mut substitution_node_id, mut substituted) = multicircuit_monad[0]
+            .builder
+            .all_nodes
+            .borrow()
+            .iter()
+            .nth(substitution_node_index % num_nodes)
+            .map(|(i, n)| (*i, n.clone()))
+            .unwrap();
+        let mut j = 0;
+        while substituted.circuit.borrow().is_zero() && j < num_nodes {
+            (substitution_node_id, substituted) = multicircuit_monad[0]
+                .builder
+                .all_nodes
+                .borrow()
+                .iter()
+                .nth((substitution_node_index + j) % num_nodes)
+                .map(|(i, n)| (*i, n.clone()))
+                .unwrap();
+            j += 1;
+        }
+        prop_assert_ne!(j, num_nodes, "no suitable nodes to substitute");
+
+        let degree_lowering_info = DegreeLoweringInfo {
+            target_degree: 2,
+            num_main_cols: main_input.len(),
+            num_aux_cols: aux_input.len(),
+        };
+        let substitution_constraint = ConstraintCircuitMonad::apply_substitution(
+            &mut multicircuit_monad,
+            degree_lowering_info,
+            substitution_node_id,
+            0,
+            0,
+        );
+
+        // extract substituted constraint
+        let CircuitExpression::BinOp(BinOp::Add, variable, neg_expression) =
+            &substitution_constraint.circuit.as_ref().borrow().expression
+        else {
+            unreachable!();
+        };
+        let extra_input =
+            match &neg_expression.as_ref().borrow().expression {
+                CircuitExpression::BinOp(BinOp::Mul, _neg_one, circuit) => circuit
+                    .borrow()
+                    .evaluate(main_input.view(), aux_input.view(), &challenges),
+                CircuitExpression::BConst(c) => -c.lift(),
+                CircuitExpression::XConst(c) => -*c,
+                _ => {
+                    unreachable!()
+                }
+            };
+        if variable.borrow().evaluates_to_base_element() {
+            main_input
+                .append(
+                    Axis(1),
+                    Array2::from_shape_vec([1, 1], vec![extra_input.coefficients[0]])
+                        .unwrap()
+                        .view(),
+                )
+                .unwrap();
+        } else {
+            aux_input
+                .append(
+                    Axis(1),
+                    Array2::from_shape_vec([1, 1], vec![extra_input])
+                        .unwrap()
+                        .view(),
+                )
+                .unwrap();
+        }
+
+        // evaluate again
```

Comment:
I think this and the next two comments are obvious from the code. If you'd like to have comments here, answering the “why” seems better to me. That said, with the top-level comment in mind, it might not be necessary to repeat.